# 5.1 - Bayes

https://en.wikipedia.org/wiki/Thomas_Bayes

![bayes](images/bayes0.jpeg)


$$A \cap B = B \cap A$$


$$P(A \cap B) = P(B \cap A)$$


$$P(A|B) = \frac{P(A \cap B)}{P(B)}$$




### Regla de Bayes

$$P(A|B) = \frac{P(B|A)P(A)}{P(B)}$$


donde:

+ $P(A|B) :=$ A posteriori
+ $P(A) :=$ A priori
+ $P(B|A) :=$ Verosimilitud (likelihood)
+ $P(B) :=$ Marginal de B


En realidad:
$$P(B) = \sum_{\Omega} P(B|A)P(A)$$


### Ejemplo: Seguro de coche segun color

**Eventos**

+ $A: =$ Accidente de coche
+ $B: =$ Color del coche (rojo, por ejemplo)


**¿Que es cada cosa?**

+ $P(A|B) :=$ ¿Prob de tener un accidente si mi coche es rojo? (lo que buscamos)
+ $P(A) :=$ Prob de accidente (proporcion de coches accidentados en todo el parque movil)
+ $P(B|A) :=$ Si ya he tenido un accidente, ¿cual es la prob de que mi coche sea rojo? (proporcion de coches rojos entre los accidentados)
+ $P(B) :=$ Prob de coche rojo (proporcion de coches rojos en todo el parque movil)




**Supongamos...**
 
+ $P(A) = 0.1$ 
+ $P(B|A) = 0.3$ 
+ $P(B) = 0.2$


**Entonces..**

In [1]:
'P(A|B) = {}'.format(0.3 * 0.1 / 0.2)  # prob de accidente si coche rojo

'P(A|B) = 0.15'

**Conclusion**

La prob de tener un accidente si mi coche es rojo es de 0.15

...pero es extensible a distribuciones..

### Interpretación diacrónica de la regla de Bayes

**Supongamos..**

+ $A: =$ H (hipótesis)
+ $B: =$ D (datos)

Entonces:


$$P(H|D) = \frac{P(D|H)P(H)}{P(D)}$$

¿Que es $P(D)$?

$$P(D) = \sum_{\Omega} P(D|H)P(H)$$


**Solucion**

Se tienen dos hipótesis, H0 y H1:

$$\frac{P(H_1|D) = \frac{P(D|H_1)P(H_1)}{P(D)}}{P(H_0|D) = \frac{P(D|H_0)P(H_0)}{P(D)}}$$


$$\frac{P(H_1|D)}{P(H_0|D)} = \frac{P(D|H_1)P(H_1)}{P(D|H_0)P(H_0)}$$


Desde aqui solo hay que obtener ese cociente y ver si es mayor o menor que 1.

### Ejercicios

#### 1. Problema de las cajas de galletas

Supongamos que se tienen dos cajas de galletas. Una de ellas contiene 30 galletas de vainilla y 10 galletas de chocolate. La otra caja contiene 20 galletas de cada. Se saca aleatoriamente una galleta y es de vainilla. ¿Cuál es la probabilidad de que la galleta venga de la primera caja?

In [2]:
import pandas as pd
import numpy as np
import pylab as plt

In [3]:
def bayes(priori, verosimilitud):  
    
    marginal=sum(np.multiply(priori, verosimilitud))
    
    posteriori=np.divide(np.multiply(priori, verosimilitud), marginal)
    
    return posteriori

In [4]:
priori=[1/2, 1/2] # cajas de galletas, uniforme porque soy ignorante

v_vainilla=[3/4, 2/4]
v_choco=[1/4, 2/4]

In [6]:
bayes(priori, v_vainilla)[0]

0.6

**0.6 => 60% de probabilidad de que venga de la caja de galletas 1.**

¿Cuál es la probabilidad de que la galleta venga de la segunda caja?

In [7]:
1 - bayes(priori, v_vainilla)[0]

0.4

In [8]:
bayes(priori, v_vainilla)[1]

0.4

**0.4 => 40% de probabilidad de que venga de la caja de galletas 2.**

¿Que ocurriría si la galleta hubiera sido de chocolate?, ¿cuales serían las probabilidades?

In [9]:
bayes(priori, v_choco)

array([0.33333333, 0.66666667])

**0.333 => 33.3% de probabilidad de que venga de la caja de galletas 1.**

**0.667 => 66.7% de probabilidad de que venga de la caja de galletas 2.**

#### 2. Problema de los M&Ms
Supongamos que tenemos dos bolsas de M&Ms:

En la **Bolsa 1**, la mezcla de colores es:
* Marron - 30%
* Amarillo - 20%
* Rojo - 20%
* Verde - 10%
* Naranja - 10%
* Mandarina - 10%


En la **Bolsa 2**, la mezcla de colores es:
* Azul - 24%
* Verde 20%
* Naranja - 16%
* Amarillo - 14%
* Rojo - 13%
* Marron - 13%

Sin saber cual es cual, se cogen aleatoriamente dos M&Ms, uno de cada bolsa. Uno es amarillo y uno verde. ¿Cuál es la probabilidad de que el amarillo venga de la bolsa 1?

**Tip:** Se necesitan cruzar las verosimilitudes.

In [10]:
priori=[1/2, 1/2]  # bolsas de m&ms

# verosimilitudes
v_marron=[0.3, 0.13]
v_amarillo=[0.2, 0.14]
v_rojo=[0.2, 0.13]
v_verde=[0.1, 0.2]
v_naranja=[0.1, 0.16]
v_mandarina=[0.1, 0]
v_azul=[0, 0.24]

In [12]:
vero_total=[v_amarillo[0]*v_verde[1], v_amarillo[1]*v_verde[0]]

vero_total

[0.04000000000000001, 0.014000000000000002]

In [14]:
bayes(priori, vero_total)[0]

0.7407407407407408

**0.7407 => 74.1% de probabilidad de que la amarilla venga de la bolsa 1.**

¿Cuál es la probabilidad de que el amarillo venga de la bolsa 2?

In [15]:
bayes(priori, vero_total)[1]

0.25925925925925924

**0.259 => 25.9% de probabilidad de que la amarilla venga de la bolsa 2.**

¿Cuales son las probabilidades del M&M verde?

Las probabilidades de la verde son las complementarias de cada una, es decir, la probabilidad de que la verde venga de la bolsa 1 es 0.259 y la probabilidad de que la verde venga de la bolsa 2 es 0.7407.

#### 3. Problema Monty Hall

[La paradoja en la que cae el 90% de la gente...](https://www.youtube.com/watch?v=1BpTBzDQuRE)


Supongamos ser el concursante en el concurso de television Let's Make a Deal. El presentador, llamado Monty Hall nos presenta tres puertas, la Puerta A, la Puerta B y la Puerta C. Él nos dice que que detrás de una de esas puertas hay un coche deportivo, y si escogemos la correcta nos lo llevaremos.

Seleccionamos la Puerta A, pero Monty lo pone un poco más interesante. Abre la Puerta B y detrás de ella no está el coche, hay una cabra. Nos pregunta si queremos cambiar de puerta o seguir con la puerta que ya habiamos elegido.

Dada está información, ¿Cuál es la probabilidad de ganar el coche si permanecemos con la Puerta A?. Y si cambiamos a la puerta C, ¿cuál sería?


In [16]:
priori=[1/3, 1/3, 1/3]

v_premio=[0, 1/2, 1]

In [17]:
bayes(priori, v_premio)

array([0.        , 0.33333333, 0.66666667])

**si no se cambia de puerta la probabilidad es 1/3**

**si se cambia de puerta la probabilidad es 2/3**

In [18]:
priori=[1/4, 1/4, 1/4, 1/4]

v_premio=[0, 1/2, 1/2, 1]

bayes(priori, v_premio)

array([0.  , 0.25, 0.25, 0.5 ])